src/trace/event.zig

```diff
@@ -0,0 +1,496 @@
+
+                var target_payload: TargetPayload = undefined;
+                if (target_payload_info == .Struct) {
+                    inline for (comptime std.meta.fieldNames(TargetPayload)) |field| {
+                        @field(target_payload, field) = @field(source_payload, field);
+                    }
+                }
+
+                return @unionInit(EventType, @tagName(tag), target_payload);
+            },
+        };
+    }
+};
+
+pub const EventTiming = union(Event.EventTag) {
+    replica_commit: struct { stage: CommitStageTag },
+    replica_aof_write,
+    replica_sync_table,
+
+    compact_beat: struct { tree: TreeEnum, level_b: u8 },
+    compact_beat_merge: struct { tree: TreeEnum, level_b: u8 },
+    compact_manifest,
+    compact_mutable: struct { tree: TreeEnum },
+    compact_mutable_suffix: struct { tree: TreeEnum },
+
+    lookup: struct { tree: TreeEnum },
+    lookup_worker: struct { tree: TreeEnum },
+
+    scan_tree: struct { tree: TreeEnum },
+    scan_tree_level: struct { tree: TreeEnum, level: u8 },
+
+    grid_read,
+    grid_write,
+
+    metrics_emit,
+
+    pub const stack_limits = std.enums.EnumArray(Event.EventTag, u32).init(.{
+        .replica_commit = enum_max(CommitStageTag),
+        .replica_aof_write = 1,
+        .replica_sync_table = 1,
+        .compact_beat = enum_max(TreeEnum) * @as(u32, constants.lsm_levels),
+        .compact_beat_merge = enum_max(TreeEnum) * @as(u32, constants.lsm_levels),
+        .compact_manifest = 1,
+        .compact_mutable = enum_max(TreeEnum),
+        .compact_mutable_suffix = enum_max(TreeEnum),
+        .lookup = enum_max(TreeEnum),
+        .lookup_worker = enum_max(TreeEnum),
+        .scan_tree = enum_max(TreeEnum),
+        .scan_tree_level = enum_max(TreeEnum) * @as(u32, constants.lsm_levels),
+        .grid_read = 1,
+        .grid_write = 1,
+        .metrics_emit = 1,
+    });
+
+    pub const stack_bases = array: {
+        var array = std.enums.EnumArray(Event.EventTag, u32).initDefault(0, .{});
+        var next: u32 = 0;
+        for (std.enums.values(Event.EventTag)) |event_type| {
+            array.set(event_type, next);
+            next += stack_limits.get(event_type);
+        }
+        break :array array;
+    };
+
+    pub const stack_count = count: {
+        var count: u32 = 0;
+        for (std.enums.values(Event.EventTag)) |event_type| {
+            count += stack_limits.get(event_type);
+        }
+        break :count count;
+    };
+
+    // Unlike with EventTracing, which neatly organizes related events underneath one another, the
+    // order here does not matter.
+    //
+    // TODO: This could be computed automatically, at least in the cases where all, or all except
+    // one of the values are enums.
+    //
+    // TODO: The enum logic is a bit wasteful. It considers the max value, so for enums that don't
+    // start at 0 (eg, the trees) it's not optimal.
+    pub fn stack(event: *const EventTiming) u32 {
```

Comment:
I accidentally nerd-sniped myself here! One neat thing we can do, to double-check the logic, is to just generate _all_ events and check that there stacks don't collide:

https://github.com/tigerbeetle/tigerbeetle/commit/fb61e0734a4796f488d0a68085b9b50e5e8c9925

Reply:
Unrelated, but is the stack still the right name here? Maybe `.index` or `.ordinal`, a we don't have real stacks for metrics?